src/test/java/com/commercetools/sync/services/impl/TaxCategoryServiceImplTest.java

```diff
@@ -203,13 +203,54 @@ void createTaxCategory_WithDraftHasNoKey_ShouldNotCreateTaxCategory() {
     @Test
     void updateTaxCategory_WithNoError_ShouldUpdateTaxCategory() {
         final TaxCategory mock = mock(TaxCategory.class);
-        when(client.execute(any())).thenReturn(CompletableFuture.completedFuture(mock));
+        final FakeClient<TaxCategoryPagedQueryResult> fakeTaxCategoryClient = new FakeClient(mock);
+        initMockService(fakeTaxCategoryClient);
         final List<UpdateAction<TaxCategory>> updateActions = Collections.singletonList(ChangeName.of("name"));
 
         final TaxCategory taxCategory = service.updateTaxCategory(mock, updateActions).toCompletableFuture().join();
 
         assertThat(taxCategory).isSameAs(mock);
-        verify(client).execute(eq(TaxCategoryUpdateCommand.of(mock, updateActions)));
+        assertThat(fakeTaxCategoryClient.isExecuted()).isTrue();
+    }
+
+    @Test
+    void fetchMatchingTaxCategoriesByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
+        // Mock sphere client to return BadGatewayException on any request.
+        final FakeClient<TaxCategory> fakeCustomObjectClient = new FakeClient<>(new BadGatewayException());
+        final List<String> errorCallBackMessages = new ArrayList<>();
+        final List<Throwable> errorCallBackExceptions = new ArrayList<>();
+
+        final TaxCategorySyncOptions spyOptions =
+            TaxCategorySyncOptionsBuilder.of(fakeCustomObjectClient)
+                                         .errorCallback((exception, oldResource, newResource, updateActions) -> {
+                                             errorCallBackMessages.add(exception.getMessage());
+                                             errorCallBackExceptions.add(exception.getCause());
+                                         })
+                                         .build();
+
+        service = new TaxCategoryServiceImpl(spyOptions);
+
+        final Set<String> keys = new HashSet<>();
+        keys.add(TAX_CATEGORY_KEY);
+
+        // test and assert
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(service.fetchMatchingTaxCategoriesByKeys(keys))
+            .failsWithin(1, TimeUnit.SECONDS)
+            .withThrowableOfType(ExecutionException.class)
+            .withCauseExactlyInstanceOf(BadGatewayException.class);
+    }
+
+    private void initMockService(@Nonnull final SphereClient fakeTaxCategoryClient) {
```

Comment:
this was better as a hook `@BeforeEach`, not sure why this is needed.

Reply:
Since we are changing the client in every method, I removed it from @beforeEach.
Example: 
final FakeClient<TaxCategory> fakeTaxCategoryClient = new FakeClient(mock);
        initMockService(fakeTaxCategoryClient);